### PR TITLE
Standardise handling of "Back" key binding

### LINKED ISSUE #2768: Player does not respect "Go back" keybind
The *Go back* keybind normally allows you to do as it says except when you're playing where your only option to escape is literally that button.
Pausing/Leaving `Player` should be possible with specified keybinds.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/59438128-player-does-not-respect-go-back-keybind?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Aergwyn**
closes #2768 

- [x] Depends on https://github.com/ppy/osu-framework/pull/1656.
- [x] Depends on https://github.com/ppy/osu-framework/pull/1654

### CODE REVIEW

**Comment by peppy** [comment]
For some reason, binding `Mouse4` to Back works in all scenarios except exiting the `GameplayMenuOverlay` when it is already visible.

**Comment by Aergwyn** [comment]
What I could find out so far.
1) Unbinding ESC will break in the same way
2) MouseButtons generally seem to not be recognized for a `GlobalAction`
3) It only happens in `GameplayMenuOverlay` but works anywhere else

**Comment by Aergwyn** [comment]
Just to mention what I wrote in Discord: Unless there are some pointers it is unlikely for me to continue this right now.

**Comment by peppy** [comment]
I'm gonna tidy this up a bit.

File: osu.Game/Graphics/UserInterface/FocusedTextBox.cs [hunk 1]
```diff
@@ -26,6 +25,7 @@ public bool HoldFocus
             {
                 focus = value;
                 if (!focus && HasFocus)
+                    //todo: replace with KillInput after ppy/osu-framework#1656 is merged.
```
Comment (smoogipoo) [inline osu.Game/Graphics/UserInterface/FocusedTextBox.cs]: Todo?

File: osu.Game/Graphics/UserInterface/OsuTextBox.cs [hunk 2]
```diff
@@ -59,5 +58,21 @@ protected override void OnFocusLost(InputState state)
         }
 
         protected override Drawable GetDrawableCharacter(char c) => new OsuSpriteText { Text = c.ToString(), TextSize = CalculatedTextSize };
+
+        public bool OnPressed(GlobalAction action)
+        {
+            if (action == GlobalAction.Back)
```
Comment (smoogipoo) [inline osu.Game/Graphics/UserInterface/OsuTextBox.cs]: I don't think this is working how you want this to work. Pressing esc in song select should delete the search text if there is any, but currently it's booting the user straight to menu.

Also, why is this in `OsuTextBox`? Surely we don't desire this functionality in every single textbox game wide? This means that pressing escape in the chat overlay would delete text? I can't actually find a case where this is being triggered correctly right now.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Graphics/UserInterface/OsuTextBox.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Graphics/UserInterface/OsuTextBox.cs]: Yeah, I intentionally moved this to standardise behaviour for the time being. If you disagree I guess we can move it back to FocusedTextBox

**Comment by smoogipoo** [comment]
How is this being tested? It still boots me out to the main menu from song select - I still haven't encountered a case where it deletes text from a focused textbox.

Also, even with esc unbound from "Go back" I can still exit settings and song select, but I can't "go back" levels in the menu button system.

**Comment by smoogipoo** [comment]
Looks ok.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**